[3D] Add support for a 2d map overlay

ptitjano opened the pull request:
## Description

This adds a new feature to the 3d map view to display an overlay of the 2d map centered around the position of the 3d camera. This allows in particular to better understand the current location of the 3d scene or to get an overview of the 2D canvas when only the 3d map view is visible.


<img width="1920" height="1048" alt="image" src="https://github.com/user-attachments/assets/4b16a4a7-6514-439d-a576-a30951c817cd" />

with camera frustum activated: 

<img width="297" height="288" alt="image" src="https://github.com/user-attachments/assets/729f0033-f809-479f-a09d-a57db1d6590f" />


## Demo 

https://github.com/user-attachments/assets/7c897a3e-1c35-48c9-b429-20199532db3b

## Update 22 january

- Added a timer to prevent too many updates
- camera frustum is now visible when the option is enabled

ptitjano committed:
qgsdebugtextureentity: Expose isDepth parameter

ptitjano committed:
qgs3dmapsettings: Add support for 2d map overlay

ptitjano committed:
qgs3dmapscene: Add support qgsmapoverlayentity

When the entity is enabled, an extent around the camera position is
computed. The greater the pitch, the wider the extent, because
the visible area is larger. This extent is then used to generate the
entity's texture.

ptitjano committed:
qgs3dmapcanvaswidget: Add support to enable 2d map overlay

The overlay can be enabled from the configuration window or from the
camera menu.

ptitjano committed:
qgsdebugtextureentity: Allow to flip texture vertically

ptitjano committed:
qgsdebugtextureentity: Add support for an offset in position

ptitjano committed:
qgsdebugtextureentity: Allow to set position from a QSize

The existing `setPosition` does not allow to set different width and
height.

ptitjano committed:
qgsframegraph: Add a debug texture renderview getter

ptitjano committed:
3d: Introduce qgsmapoverlaytexturegenerator

`QgsMapOverlayTextureGenerator` allows to generate a texture which
displays the current 2D map with an indicator which indicates where
the 3d camera is looking at and an arrow to indicate the azimuth.

The logic is similar to the existing `QgsTerrainTextureGenerator`.

This will be used by a new entity called `QgsMapOverlayEntity` in the
next commit.

ptitjano committed:
3d: Introduce qgsmapoverlayentity

An entity that displays a texture over the current 3D scene to show the current 3D location on the 2D canvas.

ptitjano committed:
3d: Rename QgsDebugTextureEntity to QgsOverlayTextureEntity

With the introduction `QgsMapOverlayEntity`, `QgsOverlayTextureEntity`
is not only used to create debug entities. Rename it accordingly.

ptitjano committed:
3d: Rename QgsDebugTextureRenderView to QgsOverlayTextureRenderView

With the introduction `QgsMapOverlayEntity`, `QgsOverlayTextureEntity`
is not only used to create debug entities. Rename it accordingly.

ptitjano commented on src/app/3d/qgs3dmapconfigwidget.cpp:
```diff
@@ -226,6 +226,10 @@ Qgs3DMapConfigWidget::Qgs3DMapConfigWidget( Qgs3DMapSettings *map, QgsMapCanvas
   groupBox->setVisible( map->sceneMode() == Qgis::SceneMode::Local );
 
+  // ==================
+  // 2D MapOverlay
```
Done.

ptitjano commented on src/app/3d/qgs3dmapcanvaswidget.cpp:
```diff
@@ -255,6 +255,13 @@ Qgs3DMapCanvasWidget::Qgs3DMapCanvasWidget( const QString &name, bool isDocked )
   } );
   mCameraMenu->addAction( mShowFrustumPolygon );
 
+  mActionShow2DMapOverlay = new QAction( tr( "Show 2D map overlay" ), this );
```
Done.

ptitjano commented on src/3d/qgs3dmapsettings.cpp:
```diff
@@ -430,6 +434,10 @@ QDomElement Qgs3DMapSettings::writeXml( QDomDocument &doc, const QgsReadWriteCon
 
+  QDomElement elemMapOverlay = doc.createElement( QStringLiteral( "map-overlay" ) );
+  elemMapOverlay.setAttribute( QStringLiteral( "enabled" ), mShowMapOverlay ? 1 : 0 );
+  elem.appendChild( elemMapOverlay );
```
Done.

nyalldawson commented:
@ptitjano 

>  I did not understand this comment. Do you think that the proposed circle + arrow design should be replaced by something else? That the styling (color, etc) of the circle and arrows should be similar to the "show visible camera area in 2d map view" option? Something else?

That's what I originally thought, but coming back to this, I think it actually should show both the camera visible area (as per "show visible camera area in 2d map view", with similar styling for consistently) AND ALSO the map point corresponding to the actual camera location. 

This would give both the camera location + direction, but also work nicely with situations when the camera is high from the terrain and looking at an angle down onto the scene.


Another unrelated thought which occurred after my initial review: I think the updates of the 2d map should be moved to a timer (a bit like the main canvas), so that they don't occur more than say every 0.5 seconds. I.e.:

- movement occurs, map starts redrawing immediately, a timer is started
- a subsequent movement occurs < 0.5 seconds after the first -- we don't immediately start the map update, but wait till the 0.5 second timeout has expired and then trigger the update

So we don't trigger updates at a frequency greater than 0.5 seconds, and at most it takes 0.5 seconds from a movement -> corresponding map update to happen.

(Note that it's not sufficient to just ignore subsequent movements which occur within 0.5 seconds of the first, as that would mean that the map location is outdated if the 3d camera stops movement immediately after this second trigger. Rather we need the rapid events to delay the final map redraw until after the timeout expires).

ptitjano commented:
> Another unrelated thought which occurred after my initial review: I think the updates of the 2d map should be moved to a timer (a bit like the main canvas), so that they don't occur more than say every 0.5 seconds. I.e.:
> 
>     * movement occurs, map starts redrawing immediately, a timer is started
> 
>     * a subsequent movement occurs < 0.5 seconds after the first -- we don't immediately start the map update, but wait till the 0.5 second timeout has expired and then trigger the update
> 
> 
> So we don't trigger updates at a frequency greater than 0.5 seconds, and at most it takes 0.5 seconds from a movement -> corresponding map update to happen.
> 
> (Note that it's not sufficient to just ignore subsequent movements which occur within 0.5 seconds of the first, as that would mean that the map location is outdated if the 3d camera stops movement immediately after this second trigger. Rather we need the rapid events to delay the final map redraw until after the timeout expires).

I have added a commit to implement it. I found 0.5 seconds to be too laggy. Did some testing and settled with 250 ms.

ptitjano committed:
qgsmapoverlayentity: Add support for frustum visualization

ptitjano committed:
qgs3dmapscene: Add support for frustum visualization of overlay

m-kuhn commented:
@ptitjano I guess this is ready for merging once the conflicts are resolved?

ptitjano committed:
qgs3dmapscene: Prevent too many map overlay updates on camera changes

This change adds a timer logic to refresh the map overlay every 250 ms
while the camera is moving. This prevents refreshing the map overlay
too many times

m-kuhn commented on src/3d/qgs3dmapscene.cpp:
```diff
@@ -526,6 +544,65 @@ void Qgs3DMapScene::onFrameTriggered( float dt )
+
+  if ( !mMap.is2DMapOverlayEnabled() )
+  {
+    if ( mMapOverlayEntity )
```
Checks for nullptr are done like this throughout the codebase, why compare to `nullptr` here?

benoitdm-oslandia replied:
OK

ptitjano commented on src/3d/qgs3dmapscene.h:
```diff
@@ -420,6 +427,10 @@ class _3D_EXPORT Qgs3DMapScene : public QObject
     int mMaxClipPlanes = 6;
 
+    //! 2d map overlay
+    QgsMapOverlayEntity *mMapOverlayEntity = nullptr;
```
Good idea, done.
